bandwidth: support ingress rate limiting using eBPF token bucket

l1b0k opened the pull request:
This commit adds ingress bandwidth rate limiting support using an eBPF token bucket implementation. This complements the existing egress bandwidth management and allows for bidirectional traffic shaping of Cilium endpoints. The QoS works on native device, same as EDT does, so traffic in the same node, will not work.

Key changes:
- Implements a token bucket rate limiter in eBPF (bpf/lib/tb.h)
- Extend cilium_throttle map to store ingress rate limits
- Extends the bandwidth manager to handle ingress bandwidth annotations
- Updates the CLI to display both ingress and egress bandwidth limits

Limitations:
- For native routing, BPF Routing is required

<!-- Description of change -->

Fixes: #15636

```release-note
bandwidth: support ingress rate limiting using eBPF token bucket
```

borkmann commented on bpf/lib/tb.h:
```diff
@@ -0,0 +1,65 @@
+	__uint(pinning, LIBBPF_PIN_BY_NAME);
+	__uint(max_entries, INGRESS_THROTTLE_MAP_SIZE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} INGRESS_THROTTLE_MAP __section_maps_btf;
```
Nice work! Just thinking out loud, do you think we could merge the two bwm maps? e.g. from the id key we use one bit to denote whether the type is ingress or egress?

Could you also add another commit extending/updating the bmw doc?

l1b0k replied:
I have extended `edt_id` and `edt_info`, making adjustments to simplify the code. I plan to submit another PR to update the documentation.

borkmann replied:
Btw, CI has some build error: 
```
CGO_ENABLED=0 go build -mod=vendor -ldflags '-X "github.com/cilium/cilium/pkg/version.ciliumVersion=1.18.0-dev f22163e86c 2025-01-22T23:19:34+08:00" -s -w -X "github.com/cilium/cilium/pkg/envoy.requiredEnvoyVersionSHA=585ba14700a2c729f024d9f0b2c694bc83a908a2" ' -tags=osusergo  -o cilium-docker
# github.com/cilium/cilium/pkg/datapath/linux/bandwidth
Error: ../pkg/datapath/linux/bandwidth/bandwidth.go:95:68: cannot use hostEpID (variable of type uint16) as bwmap.EdtIDKey value in argument to bwmap.EdtIDIndex.Query
Error: ../pkg/datapath/linux/bandwidth/bandwidth.go:99:31: not enough arguments in call to bwmap.NewEdt
	have (uint16, number, number)
	want (uint16, uint8, uint64, uint32)
Error: ../pkg/datapath/linux/bandwidth/bandwidth.go:104:39: not enough arguments in call to bwmap.NewEdt
	have (uint16, uint64, uint32)
	want (uint16, uint8, uint64, uint32)
Error: ../pkg/datapath/linux/bandwidth/bandwidth.go:113:70: cannot use epID (variable of type uint16) as bwmap.EdtIDKey value in argument to bwmap.EdtIDIndex.Query
make[1]: *** [Makefile:22: cilium-agent] Error 1
make[1]: Leaving directory '/go/src/github.com/cilium/cilium/daemon'
make: *** [Makefile:81: daemon] Error 2
make: *** Waiting for unfinished jobs....
```

l1b0k committed:
bpf/bandwidth: Implement token bucket algorithm for bandwidth management

Key changes:
- Implements a token bucket rate limiter in eBPF (bpf/lib/tb.h)
- Extend the edt_info struct, add the tokens field

The QoS works on native device, same as EDT does, so traffic in the same node, will not work.

Limitations:
- For native routing, BPF Routing is required

Signed-off-by: l1b0k <libokang.dev@gmail.com>

l1b0k committed:
bandwidth: support ingress rate limiting using eBPF token bucket

This commit adds ingress bandwidth rate limiting support using an eBPF token bucket
implementation. This complements the existing egress bandwidth management and allows
for bidirectional traffic shaping of Cilium endpoints.

Key changes:
- Extends the bandwidth manager to handle ingress bandwidth annotations
- Updates the CLI to display both ingress and egress bandwidth limits

Signed-off-by: l1b0k <libokang.dev@gmail.com>

borkmann commented on bpf/bpf_overlay.c:
```diff
@@ -145,9 +146,13 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
+	if (ep && !(ep->flags & ENDPOINT_MASK_HOST_DELIVERY)) {
+		ret = accept(ctx, ep->lxc_id);
+		if (IS_ERR(ret))
+			return ret;
```
Looks great! Small nit: Could we move this directly into ipv{4,6}_local_delivery()? I presume this needs to exclude local Pod<>Pod then.

borkmann commented on bpf/lib/common.h:
```diff
@@ -342,15 +342,23 @@ struct endpoint_info {
 struct edt_id {
 	__u64		id;
+	__u8        direction;
+	__u8        pad[7];
```
tiny nit: whitespace got messed up, should be tab

borkmann commented on bpf/lib/common.h:
```diff
@@ -342,15 +342,23 @@ struct endpoint_info {
+		__u64  t_horizon_drop;
+		__u64  tokens;
+	};
+    __u32		prio;
```
ditto

borkmann reviewed (approved):
modulo small nits, LGTM from bpf side, thanks!

nathanjsweet reviewed (approved):
Looks good for K8s, CLI, and API.

l1b0k committed:
doc: add ingress for bandwidth

Signed-off-by: l1b0k <libokang.dev@gmail.com>

youngnick reviewed (approved):
k8s related changes are minor and LGTM

jrajahalme commented on bpf/lib/common.h:
```diff
@@ -342,14 +342,22 @@ struct endpoint_info {
 struct edt_id {
 	__u64		id;
+	__u8		direction;
+	__u8		pad[7];
```
Given that the `id` field is always(?) set to a 32-bit value, it would be better redefine this as:
```
struct edt_id {
	__u32		id;
	__u8		direction;
	__u8		pad[3];
};
```
This would also keep the size of `edt_id` intact. This is nice since many map implementations may have optimized hashing of 32 or 64-bit keys (not sure if applies to bpf maps, though).

Corresponding change would need to be done on Go `type EdtId` as well.

Looks like in userspace the `id` is only 16 bits, for all this would fit into a 32-bit key in kernel as well.

l1b0k replied:
Good call, I've made that change.

jrajahalme reviewed (changes requested):
I'd rather not double the bpf map key size, when that is not necessary.

qmonnet commented on bpf/lib/edt.h:
```diff
@@ -44,7 +44,7 @@ static __always_inline int
 {
 	__u64 delay, now, t, t_next;
-	struct edt_id aggregate;
+	struct edt_id aggregate = {0};
```
Style nit: Most other locations do `= {}` rather than `= {0}`

qmonnet commented on bpf/lib/tb.h:
Would you consider renaming this file to make the name more explicit, please? `token_bucket.h` is not _that_ long and would make it easier to understand what it's about.

l1b0k replied:
sure

qmonnet reviewed (changes requested):
Looks good thank you, just one comment on the file name.

qmonnet reviewed (approved):
Looks good, thank you!

hemanthmalla reviewed (approved):
LGTM. Thank you.
